module/zfs/zap_micro.c

```diff
@@ -149,43 +149,104 @@ zap_normalize(zap_t *zap, const char *name, char *namenorm, int normflags)
 boolean_t
 zap_match(zap_name_t *zn, const char *matchname)
 {
+	boolean_t res = B_FALSE;
 	ASSERT(!(zap_getflags(zn->zn_zap) & ZAP_FLAG_UINT64_KEY));
 
 	if (zn->zn_matchtype & MT_NORMALIZE) {
-		char norm[ZAP_MAXNAMELEN];
+		size_t namelen = zn->zn_normbuf_len;
+		char normbuf[ZAP_MAXNAMELEN];
+		char *norm = normbuf;
 
-		if (zap_normalize(zn->zn_zap, matchname, norm,
-		    zn->zn_normflags) != 0)
-			return (B_FALSE);
+		/*
+		 * Cannot allocate this on-stack as it exceed the stack-limit of
+		 * 1024.
+		 */
+		if (namelen > ZAP_MAXNAMELEN)
+			norm = kmem_alloc(namelen, KM_SLEEP);
```

Comment:
Can we get rid of normbuf[] all together, or is there there a benefit to using the stack when possible?

Reply:
Just thinking out loud - I'm curious if it would be worth it to do:
```diff
- norm = kmem_alloc(namelen, KM_SLEEP);
+ norm = kmem_cache_alloc(zap_name_long, KM_SLEEP);
```
You would waste some memory, but allocations might be faster.  I dunno if zap_match() is a critical path or not though.

Reply:
> Can we get rid of normbuf[] all together, or is there there a benefit to using the stack when possible?

I'm ok with removing. I think it only matters if case insensitive or some form of utf8 normalization is inuse.